File: apps/agni-server/src/core/interactions/notification/pushNotification.ts
import { EventContent, IEventListener } from "@core/adapters/event";
import { IUsecase } from "../interfaces";
import Repository from "@core/adapters/repository";
import Notification from "@core/domains/entities/notification";
import { GetUID } from "@core/adapters/libs";
import { CreatedDto } from "@core/dto/base";

export type RequestPushNotification = {
    title: string
    content: string
    dateTime: Date
}

export class PushNotificationUseCase implements IUsecase<RequestPushNotification, CreatedDto>, IEventListener  { 
    private notificationRepo: Repository<Notification>

    constructor(notificationRepo: Repository<Notification>) {
        this.notificationRepo = notificationRepo
    }

    async execute(request: RequestPushNotification): Promise<CreatedDto> {
        const notification = new Notification(GetUID(), request.title, request.content, false, request.dateTime)
        await this.notificationRepo.create(notification)
        return {newId: notification.getId()} 
    }

    update(event: EventContent): void {
        this.execute({
            title: event.title,
            content: event.content,
            dateTime: new Date() 
        })
    }
}

File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/core/usecases/notifications/PushNotification.kt
package dev.auguste.agni_api.core.usecases.notifications

import dev.auguste.agni_api.core.adapters.events.listeners.INotificationEventListener
import dev.auguste.agni_api.core.adapters.events.contents.NotificationEventContent
import dev.auguste.agni_api.core.adapters.repositories.IRepository
import dev.auguste.agni_api.core.entities.Notification
import dev.auguste.agni_api.core.usecases.CreatedOutput
import dev.auguste.agni_api.core.usecases.interfaces.IUseCase
import dev.auguste.agni_api.core.usecases.notifications.dto.PushNotificationInput

class PushNotification(private val notificationRepo: IRepository<Notification>): IUseCase<PushNotificationInput, CreatedOutput>, INotificationEventListener {
    private var event: NotificationEventContent? = null

    override fun execAsync(input: PushNotificationInput): CreatedOutput {
        val newNotification = Notification(
            title = input.title,
            content = input.content
        )

        notificationRepo.create(newNotification)

        return CreatedOutput(newNotification.id)
    }

    override fun update() {
        event?.let {
            execAsync(PushNotificationInput(
                title = it.title,
                content = it.message
            ))
        }
        // free
        event = null
    }

    override fun serve(content: NotificationEventContent) {
        event = content
    }
}